File: app/src/main/java/com/example/puyo_base_simulator/ui/home/StackWithButton.java
package com.example.puyo_base_simulator.ui.home;

import java.util.Stack;

public class StackWithButton<T> extends Stack<T> {
    ButtonUpdateFunction enableFun;
    ButtonUpdateFunction disableFun;

    StackWithButton(ButtonUpdateFunction enableFun, ButtonUpdateFunction disableFun) {
        super();
        this.enableFun = enableFun;
        this.disableFun = disableFun;
        disableFun.func();
    }

    @Override
    public T push(T elm) {
        enableFun.func();
        return super.push(elm);
    }

    @Override
    public T pop() {
        T elm =  super.pop();
        if (super.isEmpty()) disableFun.func();
        return elm;
    }

    @Override
    public void clear() {
        super.clear();
        disableFun.func();
    }
}

File: app/src/main/java/com/example/puyo_base_simulator/ui/home/StackWithButton.kt
package com.example.puyo_base_simulator.ui.home

import java.util.*

class StackWithButton<T> internal constructor(private val enableFun: () -> Unit,
                                              private val disableFun: () -> Unit) : Stack<T>() {
    override fun push(elm: T): T {
        enableFun.invoke()
        return super.push(elm)
    }

    override fun pop(): T {
        val elm = super.pop()
        if (super.isEmpty()) disableFun.invoke()
        return elm
    }

    override fun clear() {
        super.clear()
        disableFun.invoke()
    }

    init {
        disableFun.invoke()
    }
}